src/Compilers/Core/Portable/SourceGeneration/RunResults.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Text;
+
+#nullable enable
+
+namespace Microsoft.CodeAnalysis
+{
+    public readonly struct GeneratorDriverRunResult
+    {
+        internal GeneratorDriverRunResult(ImmutableArray<GeneratorRunResult> results, ImmutableArray<SyntaxTree> trees, ImmutableArray<Diagnostic> diagnostics)
+        {
+            this.Results = results;
+            this.SyntaxTrees = trees;
+            this.Diagnostics = diagnostics;
+        }
+
+        public ImmutableArray<GeneratorRunResult> Results { get; }
+
+        public ImmutableArray<Diagnostic> Diagnostics { get; }
+
+        public ImmutableArray<SyntaxTree> SyntaxTrees { get; }
+    }
+
+    public readonly struct GeneratorRunResult
+    {
+        internal GeneratorRunResult(ISourceGenerator generator, ImmutableArray<GeneratedSourceResult> generatedTrees, ImmutableArray<Diagnostic> diagnostics, Exception? exception)
+        {
+            this.Generator = generator;
+            this.GeneratedTrees = generatedTrees;
+            this.Diagnostics = diagnostics;
+            this.Exception = exception;
+        }
+
+        public ISourceGenerator Generator { get; }
+
+        public ImmutableArray<GeneratedSourceResult> GeneratedTrees { get; }
+
+        public ImmutableArray<Diagnostic> Diagnostics { get; }
+
+        public Exception? Exception { get; }
+
+        public bool Succeeded => Exception is null;
```

Comment:
Why do we consider a result succeeded when there are error diagnostics?

Reply:
From the point of view of the generator driver, a generation pass is successful even if the generator added error diagnostics. I'm not convinced this is necessary to expose explicitly though, so I'm going to remove.

---
In reply to: [460214544](https://github.com/dotnet/roslyn/pull/46296#discussion_r460214544) [](ancestors = 460214544)